src/Microsoft.OData.Core/MultipartMixed/ODataMultipartMixedBatchWriter.cs

```diff
@@ -0,0 +1,569 @@
+        {
+            this.WritePendingMessageData(true);
+
+            // In responses we don't need to use our batch URL resolver, since there are no cross referencing URLs
+            // so use the URL resolver from the batch message instead.
+            this.CurrentOperationResponseMessage = ODataBatchOperationResponseMessage.CreateWriteMessage(
+                this.RawOutputContext.OutputStream,
+                /*operationListener*/ this,
+                this.PayloadUriConverter.BatchMessagePayloadUriConverter,
+                this.Container);
+            this.SetState(BatchWriterState.OperationCreated);
+
+            Debug.Assert(this.CurrentOperationContentId == null, "The Content-ID header is only supported in request messages.");
+
+            // write the operation's start boundary string
+            this.WriteStartBoundaryForOperation();
+
+            // write the headers and request separator line
+            ODataBatchWriterUtils.WriteResponsePreamble(this.RawOutputContext.TextWriter, changeSetBoundary != null, contentId);
+
+            return this.CurrentOperationResponseMessage;
+        }
+
+        /// <summary>
+        /// Writes all the pending headers and prepares the writer to write a content of the operation.
+        /// </summary>
+        protected override void StartBatchOperationContent()
+        {
+            Debug.Assert(this.CurrentOperationMessage != null, "Expected non-null operation message!");
+            Debug.Assert(this.RawOutputContext.TextWriter != null, "Must have a batch writer!");
+
+            // write the pending headers (if any)
+            this.WritePendingMessageData(false);
+
+            // flush the text writer to make sure all buffers of the text writer
+            // are flushed to the underlying async stream
+            this.RawOutputContext.TextWriter.Flush();
+        }
+
+        /// <summary>
+        /// Disposes the batch writer and set the 'OperationStreamRequested' batch writer state;
+        /// called after the flush operation(s) have completed.
+        /// </summary>
+        protected override void DisposeBatchWriterAndSetContentStreamRequestedState()
+        {
+            this.RawOutputContext.CloseWriter();
+
+            this.SetState(BatchWriterState.OperationStreamRequested);
+        }
+
+        /// <summary>
+        /// Verifies that calling CreateOperationRequestMessage is valid.
+        /// </summary>
+        /// <param name="synchronousCall">true if the call is to be synchronous; false otherwise.</param>
+        /// <param name="method">The Http method to be used for this request operation.</param>
+        /// <param name="uri">The Uri to be used for this request operation.</param>
+        /// <param name="contentId">The Content-ID value to write in ChangeSet head.</param>
+        protected override void VerifyCanCreateOperationRequestMessage(bool synchronousCall, string method, Uri uri, string contentId)

```

Comment:
>VerifyCanCreateOperationRequestMessage [](start = 32, length = 38)

VerifyCanCreateOperationRequestMessage should be private in the base class, and include the logic for CanCreateOperationRequestMessageVerifierCommon. The logic in VerifyCanCreateOperationRequestMessageAgainstChangeSetBoundary should be in CreateOperationRequestMessageImplementation.